### ISSUE TITLE
bug: array_access panic but array_range_access works as expected

### ISSUE THREAD

**Issue by lmatz** [risingwavelabs/risingwave#9494]
### Describe the bug

```
dev=> select message from web_logs order by id limit 1;
                              message                               
--------------------------------------------------------------------
 Apr 24 14:53:18 ip-10-0-220-106 web: > meta-search-web@0.1.0 start
(1 row)

dev=> select string_to_array(message,' ') from web_logs order by id limit 1;
                           string_to_array                            
----------------------------------------------------------------------
 {Apr,24,14:53:18,ip-10-0-220-106,web:,>,meta-search-web@0.1.0,start}
(1 row)

dev=> select string_to_array(message,' ')[1:7] from web_logs order by id limit 1;
                            ?column?                            
----------------------------------------------------------------
 {Apr,24,14:53:18,ip-10-0-220-106,web:,>,meta-search-web@0.1.0}
(1 row)

dev=> select string_to_array(message,' ')[1:8] from web_logs order by id limit 1;
                               ?column?                               
----------------------------------------------------------------------
 {Apr,24,14:53:18,ip-10-0-220-106,web:,>,meta-search-web@0.1.0,start}
(1 row)

dev=> select string_to_array(message,' ')[1:9] from web_logs order by id limit 1;
                               ?column?                               
----------------------------------------------------------------------
 {Apr,24,14:53:18,ip-10-0-220-106,web:,>,meta-search-web@0.1.0,start}
(1 row)

dev=> select string_to_array(message,' ')[7] from web_logs order by id limit 1;
ERROR:  QueryError: internal error: error reading a body from connection: stream closed because of a broken pipe

dev=> select string_to_array(message,' ')[7:7] from web_logs order by id limit 1;
        ?column?         
-------------------------
 {meta-search-web@0.1.0}
(1 row)
```

LOG:
```
thread 'risingwave-batch-tasks' panicked at 'assertion failed: idx < self.len()', src/common/src/buffer/bitmap.rs:309:9
   0:     0x5579f767d56f - std::backtrace_rs::backtrace::libunwind::trace::h2ae08ac77e721453
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x5579f767d56f - std::backtrace_rs::backtrace::trace_unsynchronized::hb7d3015d68fe20ce
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5579f767d56f - std::sys_common::backtrace::_print_fmt::h60e26e0e8c18d6ce
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x5579f767d56f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h3b0d15c01f4c6ce1
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x5579f230020f - core::fmt::write::h53efc8feba3ef479
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/fmt/mod.rs:1254:17
   5:     0x5579f7677d93 - std::io::Write::write_fmt::h36688c4bf2bf4993
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/io/mod.rs:1698:15
   6:     0x5579f767d325 - std::sys_common::backtrace::_print::hebc179fe7d730070
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x5579f767d325 - std::sys_common::backtrace::print::h283655fa821c834e
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x5579f767ea4f - std::panicking::default_hook::{{closure}}::h169b3603060ea40e
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:269:22
   9:     0x5579f767e80e - std::panicking::default_hook::ha74066679145a3c4
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:288:9
  10:     0x5579f7048fdd - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h98ae094da7503a5a
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/alloc/src/boxed.rs:1990:9
  11:     0x5579f7048fdd - risingwave_rt::set_panic_hook::{{closure}}::h9c4efecfafa0adea
                               at /risingwave/src/utils/runtime/src/lib.rs:115:9
  12:     0x5579f7048fdd - std::panicking::update_hook::{{closure}}::hbc02f53a682369d3
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:234:47
  13:     0x5579f767f170 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hfa989ed38d86b9d9
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/alloc/src/boxed.rs:1990:9
  14:     0x5579f767f170 - std::panicking::rust_panic_with_hook::h252878fc82e8de0b
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:694:13
  15:     0x5579f767eeb5 - std::panicking::begin_panic_handler::{{closure}}::h38fb6f7c142dfe90
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:579:13
  16:     0x5579f767d986 - std::sys_common::backtrace::__rust_end_short_backtrace::ha14925efe8e462d4
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/sys_common/backtrace.rs:150:18
  17:     0x5579f767ec72 - rust_begin_unwind
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/std/src/panicking.rs:577:5
  18:     0x5579f22fcee3 - core::panicking::panic_fmt::h15fa664bfbb8bffe
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/panicking.rs:67:14
  19:     0x5579f22fcf7d - core::panicking::panic::ha9c8ce6950c50fe3
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/panicking.rs:117:5
  20:     0x5579f32fa03b - risingwave_common::array::ArrayImpl::value_at::h960cbe59efbd66a7
                               at /risingwave/src/common/src/array/mod.rs:608:60
21:     0x5579f33ac760 - risingwave_common::array::list_array::ListRef::value_at::h2d782413f5f60fb5
                               at /risingwave/src/common/src/array/list_array.rs:413:24
  22:     0x5579f529269e - risingwave_expr::vector_op::array_access::array_access::h228e374e64894668
                               at /risingwave/src/expr/src/vector_op/array_access.rs:27:20
  23:     0x5579f4dce51d - risingwave_expr::vector_op::array_access::array_access_list_int32_varchar::{{closure}}::{{closure}}::h97ad786975767de7
                               at /risingwave/src/expr/src/vector_op/array_access.rs:21:1
  24:     0x5579f4dce51d - <risingwave_expr::expr::template::BinaryNullableExpression<IA1,IA2,OA,F> as risingwave_expr::expr::Expression>::eval_v2::{{closure}}::hede1db9e6f9202d5
                               at /risingwave/src/expr/src/expr/template.rs:374:1
  25:     0x5579f4eb3ade - <core::pin::Pin<P> as core::future::future::Future>::poll::h55885714590b3fef
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/future/future.rs:125:9
  26:     0x5579f4eb3ade - risingwave_expr::expr::Expression::eval::{{closure}}::h624b344bc68ea28d
                               at /risingwave/src/expr/src/expr/mod.rs:111:40
  27:     0x5579f2f8f506 - <core::pin::Pin<P> as core::future::future::Future>::poll::h6b9453694404b3ea
                               at /rustc/28a29282f6dde2e4aba6e1e4cfea5c9430a00217/library/core/src/future/future.rs:125:9
  28:     0x5579f2f8f506 - risingwave_batch::executor::project::ProjectExecutor::do_execute::{{closure}}::{{closure}}::{{closure}}::{{closure}}::hc9f8e5920312bbb9
                               at /risingwave/src/batch/src/executor/project.rs:66:81
```

### To Reproduce

~~Copy paste that message, insert into some table CANNOT reproduce it.~~
~~The message is read from kinesis.~~

```
create table web_logs (id int, message varchar);
insert into web_logs values (7, 'Apr 24 14:53:18 ip-10-0-220-106 web: > meta-search-web@0.1.0 start'), (9, 'aa bb');
select string_to_array(message,' ')[7] from web_logs order by id limit 1;
```

### Expected behavior

```
 meta-search-web@0.1.0
```

### Additional context

Deploy `v0.19.0-rc3` image via https://www.risingwave.dev/docs/current/risingwave-kubernetes/

**Comment by xiangjinwu** [comment]
Similar panic trace can be obtained by the following query:
```
with t(v) as (values (array[1, 2]), (array[3, 4, 5])) select v[4] from t;
```

Besides fixing this repro query, 2 more things remain unanswered:
* > Copy paste that message, insert into some table CANNOT reproduce it.
* There seems to be no index out of range error in the original `string_to_array(message,' ')[7]` expression.

**Comment by xiangjinwu** [comment]
This issue is already fixed on `main` by a refactor #9532. The old buggy and inefficient implementation was rewritten.
